src/Nethermind/Nethermind.AuRa/AuRaBlockFinalizationManager.cs

```diff
@@ -0,0 +1,269 @@
+//  Copyright (c) 2018 Demerzel Solutions Limited
+//  This file is part of the Nethermind library.
+// 
+//  The Nethermind library is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU Lesser General Public License as published by
+//  the Free Software Foundation, either version 3 of the License, or
+//  (at your option) any later version.
+// 
+//  The Nethermind library is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+//  GNU Lesser General Public License for more details.
+// 
+//  You should have received a copy of the GNU Lesser General Public License
+//  along with the Nethermind. If not, see <http://www.gnu.org/licenses/>.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using Nethermind.AuRa.Validators;
+using Nethermind.Blockchain;
+using Nethermind.Core;
+using Nethermind.Core.Crypto;
+using Nethermind.Logging;
+using Nethermind.Store;
+using Nethermind.Store.Repositories;
+
+namespace Nethermind.AuRa
+{
+    public class AuRaBlockFinalizationManager : IBlockFinalizationManager
+    {
+        private static readonly List<BlockHeader> Empty = new List<BlockHeader>();
+        private readonly IBlockTree _blockTree;
+        private readonly IBlockInfoRepository _blockInfoRepository;
+        private readonly IAuRaValidator _auRaValidator;
+        private readonly ILogger _logger;
+        private readonly IBlockProcessor _blockProcessor;
+        private long _lastFinalizedBlockLevel = -1L;
+        private Keccak _lastFinalizingBLockHash = Keccak.EmptyTreeHash;
+        private ValidationStampCollection _lastFinalizingBLockValidators = new ValidationStampCollection();
+
+        public AuRaBlockFinalizationManager(IBlockTree blockTree, IBlockInfoRepository blockInfoRepository, IBlockProcessor blockProcessor, IAuRaValidator auRaValidator, ILogManager logManager)
+        {
+            _blockTree = blockTree ?? throw new ArgumentNullException(nameof(blockTree));
+            _blockInfoRepository = blockInfoRepository ?? throw new ArgumentNullException(nameof(blockInfoRepository));
+            _auRaValidator = auRaValidator ?? throw new ArgumentNullException(nameof(auRaValidator));
+            _logger = logManager?.GetClassLogger() ?? throw new ArgumentNullException(nameof(logManager));
+            _blockProcessor = blockProcessor ?? throw new ArgumentNullException(nameof(blockProcessor));
+            _blockProcessor.BlockProcessed += OnBlockProcessed;
+            InitLastFinalizedBLock();
+            auRaValidator.SetFinalizationManager(this);
```

Comment:
can solve via events? these two-way dependencies often lead to errors

Reply:
Tried to avoid it as much as I can but didn't find a way. I am happy to discuss it, maybe you will have some good idea.

Reply:
then maybe invert initialization order and just register to the event in the constructor?